Add namespaces

alexdavey opened the pull request:
# Description

This PR adds "namespaces", a directory-like hierarchical structure that can be used to group datasets together. For example, creating a dataset with id `my_namespace/cartpole-test-v0` will create a namespace `my_namespace`, containing the CartPole dataset as well with all other datasets starting with "my_namespace/". Namespaces can be arbitrarily nested, and can be assigned metadata which can be used describe the datasets it contains (or any other auxiliary data). These features should help organise the Minari database as it grows, and particularly as third-party datasets are added.

Dataset ids without a "/" are considered to be in the "top-level" namespace, thus this is a non-breaking change.

Internally, namespaces affect three things:
- The first part of the `dataset_id`, e.g. `my_namespace` in `my_namespace/cartpole-test-v0`.
- The directory in which the dataset is stored. E.g. `my_namespace/cartpole-test-v0` would be stored in the subdirectory `my_namespace` in the Minari datasets directory.
- Namespace metadata files `namespace_metadata.json`, which distinguish a namespace from an ordinary folder, and can hold optional metadata.

Functionality is provided to create, delete and update local namespaces in `minari.namespaces`. Namespaces can also be uploaded to the remote Farama server. Since this introduces a high potential for data conflicts, the interface is rather restrictive. Once non-empty namespace metadata is uploaded to the remote it cannot be overwritten and would need to be modified manually on the server. Namespace data is not synced between the local and remote database, except when uploading a dataset with a namespace that does not exist in the remote. Metadata can be synced manually using `upload_namespace()` and `download_namespace_data()`.

To provide some test coverage of the remote functionality, I have created a test GCP bucket [minari-datasets-test](https://console.cloud.google.com/storage/browser/minari-datasets-test) which holds some dummy namespaced datasets. I would be happy to move this over to an official Farama server, or alternatively remove the tests in `test_namespace_remote.py` until there are some namespaced datasets in the main GCP bucket.

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have run `pytest -v` and no errors are present.
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I solved any possible warnings that `pytest -v` has generated that are related to my code to the best of my knowledge.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

younik reviewed (changes requested):
Thank you for this massive work @alexdavey!
It looks good, I had some comments on the code.

Moreover, I want to write down reasoning on high-level implementation choices.

1) Shall we merge the data group creation with dataset combine? What are the differences?
It would be nice to download all datasets in a group and even load all datasets of a group. The main obstacle is that now we suppose all datasets to have the same observation and action space. With a dataset group, they can differ. Moreover, we should implement a soft-link between datasets (we will need to do it in the near future anyway).
With the increasing interest in large foundation models, I think having a big group of "soft-linked" datasets is nice to have. However, the user will need to implement processing mechanisms specific to Gym spaces, so leaving it to the user can be anyway the solution.

2) Shall we have the env name and dataset name as a namespace as well?
Again what are the differences between these two features? The main difference I see is in the doc, but we can treat the cases separately in it.

3) How should we change the `minari list` output? One of the reasons for these features was to improve the readability (and velocity) of it when we increase the number of datasets. So an idea is to simply show the dataset groups. Then the user can use `minari list namespace` to show groups/datasets inside that namespace.

younik commented on docs/content/dataset_standards.md:
```diff
@@ -55,6 +59,10 @@ For each episode group the default metadata attributes are:
+## Namespace metadata
+
+Namespaces can have metadata associated to them, to describe the group of datasets that they contain. Arbitrary JSON-serializable metadata is supported, and is stored in the file `namespace_metadata.json` in the appropriate namespace directory.
+
```
Oh I saw it later, is to easily differentiate namespaces from datasets. Anyway, see point 2 for discussion

alexdavey commented:
Thank you for the review @younik! I believe I have addressed most of the review comments.

I also modified `download_dataset()` so that it also creates an empty namespace if one does not already exist. On the remote, a namespace is distinguished from an ordinary folder by the presence of a (possibly empty) `namespace_metadata.json`. So this change makes the local behaviour more consistent with the remote.

As for the high-level discussion, here are my thoughts:

> 1. Shall we merge the data group creation with dataset combine? What are the differences?  
> It would be nice to download all datasets in a group and even load all datasets of a group. The main obstacle is that now we suppose all datasets to have the same observation and action space. With a dataset group, they can differ. Moreover, we should implement a soft-link between datasets (we will need to do it in the near future anyway).  
    With the increasing interest in large foundation models, I think having a big group of "soft-linked" datasets is nice to have. However, the user will need to implement processing mechanisms specific to Gym spaces, so leaving it to the user can be anyway the solution.

I would consider dataset namespaces a purely organisational tool, that makes no assumption about the datasets within, unlike combined datasets. Namespaces are really just an abstraction over directories.

Although the functionality is not available at the moment, it would be straightforward to extend the namespace API to include moving datasets in and out of namespaces, as well as downloading an entire namespace with `minari download my_namespace`. It's not clear to me how loading an entire namespace would work in any way that improves over manually loading a list of datasets one by one.

Soft-linked datasets are are a nice idea. However, it seems like a fundamentally different concept to that of a namespace. In addition to the observation/action space issue mentioned, soft-linked datasets could potentially access multiple namespaces.

    
> 2. Shall we have the env name and dataset name  as a namespace as well?  
    Again what are the differences between these two features? The main difference I see is in the doc, but we can treat the cases separately in it.

That’s a good idea, I guess the most flexible approach would be to merge the `env_name` into the `dataset name`. This would preserve backwards compatibility with existing namespaces, and would allow people to create a namespace e.g. `"door"` to hold `["door/human-v0", "door/cloned-v0", ...]` or store them without a namespace `["door-human-v0", "door-cloned-v0", ...]`. What do you think?

    
> 3. How should we change the `minari list` output? One of the reasons for these features was to improve the readability (and velocity) of it when we increase the number of datasets. So an idea is to simply show the dataset groups. Then the user can use `minari list namespace` to show groups/datasets inside that namespace.

In this PR, the modification to `minari list` is minimal. For example:

<img width="944" alt="Screenshot 2024-07-13 at 17 35 50" src="https://github.com/user-attachments/assets/712b5d9d-c0e0-4d4a-8e0d-c82f740b0c79">


With the ~30 datasets in the remote, this works fairly well. Adding `minari list namespace` is a good idea, but I think it's somewhat difficult to predict how we should modify it further until there are significantly more datasets (with namespaces) in the remote server.

younik commented on minari/storage/hosting.py:
```diff
@@ -188,20 +209,30 @@ def list_remote_datasets(
                 dataset_id = metadata["dataset_id"]
-                env_name, dataset_name, version = parse_dataset_id(dataset_id)
-                dataset = f"{env_name}-{dataset_name}"
+                dataset_location = os.path.dirname(os.path.dirname(blob.name))
+
+                if dataset_id != dataset_location:
+                    raise ValueError(
+                        f"Namespace location '{dataset_location}' does not match id '{dataset_id}'."
+                    )
```
No harm in this check; however, I don't know if it is a good idea to force the dataset_id to correspond to dataset_location.

younik commented on minari/storage/local.py:
```diff
@@ -14,6 +18,19 @@
+def dataset_id_sort_key(dataset_id: str) -> Tuple[str, Union[str, None], str, int]:
+    """Key for sorting dataset ids first by namespace, and then alphabetically."""
+    attrs = parse_dataset_id(dataset_id)
+    return ("" if attrs[0] is None else attrs[0],) + attrs[1:]
```
Okay, so if two datasets have the same namespace (e.g. "") and the env is None, we will get an error when try to sort, isn't it? Because sort try to compare None

younik commented on tests/test_namespace.py:
```diff
@@ -0,0 +1,122 @@
+    create_namespace(namespace, description="a new definition", overwrite=True)
+    assert get_namespace_metadata(namespace) == {"description": "a new definition"}
+
+    update_namespace(namespace, description="a third definition")
```
yes, I understand. What do you think about having description and then **kwargs (or **metadata) for the rest of metadata?

alexdavey replied:
That's a good suggestion, thanks!

younik commented:
Thank you @alexdavey!


Regarding 1), yes it makes sense, let's leave as it is.

About 2), we can actually modify things in this release as we moved metadata to the JSON file (previous release read metadata from the hdf5 file). So we can even change the name of the datasets and the previous version should work without problem. So, we can change the datasets name to something like "door/human/v2". Or, we could use the "-" for nested namespaces. 

Yes, let's leave 3) for the future.
